File: leetcode-cn/23.合并k个升序链表.py
#
# @lc app=leetcode.cn id=23 lang=python3
#
# [23] 合并K个升序链表
#

# @lc code=start
# Definition for singly-linked list.
from typing import List


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        p1 = l1
        p2 = l2

        ans = ListNode()
        q = ans

        while p1 is not None and p2 is not None:
            if p1.val <= p2.val:
                q.next = p1
                q = q.next
                p1 = p1.next
            else:
                q.next = p2
                q = q.next
                p2 = p2.next

        if p1:
            q.next = p1

        if p2:
            q.next = p2

        return ans.next

    def merge(self, lists, left, right):
        if left == right:
            return lists[left]
        if left > right:
            return None

        mid = (left + right) >> 1
        left_list = self.merge(lists, left, mid)
        right_list = self.merge(lists, mid + 1, right)

        return self.mergeTwoLists(left_list, right_list)

    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        return self.merge(lists, 0, len(lists) - 1)


# @lc code=end

File: leetcode-cn/C++/23.合并k个升序链表.cpp
/*
 * @lc app=leetcode.cn id=23 lang=cpp
 *
 * [23] 合并K个升序链表
 */

// @lc code=start
#include <vector>
using namespace std;

// struct ListNode {
//     int val;
//     ListNode* next;
//     ListNode()
//         : val(0)
//         , next(nullptr)
//     {
//     }
//     ListNode(int x)
//         : val(x)
//         , next(nullptr)
//     {
//     }
//     ListNode(int x, ListNode* next)
//         : val(x)
//         , next(next)
//     {
//     }
// };

class Solution {
public:
    ListNode* mergeTwoList(ListNode* a, ListNode* b)
    {
        if (!a) {
            return b;
        }
        if (!b) {
            return a;
        }

        ListNode* ans = new ListNode(0);
        ListNode* cur = ans;

        while (a && b) {
            if (a->val <= b->val) {
                cur->next = a;
                a = a->next;
            } else {
                cur->next = b;
                b = b->next;
            }
            cur = cur->next;
        }
        if (a) {
            cur->next = a;
        } else {
            cur->next = b;
        }
        return ans->next;
    }

    ListNode* merge(vector<ListNode*>& lists, int left, int right)
    {
        if (left == right) {
            return lists[left];
        }
        if (left > right) {
            return nullptr;
        }
        int mid = (left + right) / 2;
        ListNode* merge_left = merge(lists, left, mid);
        ListNode* merge_right = merge(lists, mid + 1, right);
        return mergeTwoList(merge_left, merge_right);
    }

    ListNode* mergeKLists(vector<ListNode*>& lists)
    {
        return merge(lists, 0, lists.size() - 1);
    }
};
// @lc code=end
